File: Editorial/132_Palindrome_Partitioning_II/TypeScript/6_Expand_Around_the_Center.ts
function minCut(s: string): number {
    let cutsDp: number[] = new Array(s.length).fill(0);
    for (let i = 1; i < s.length; i++) {
        cutsDp[i] = i;
    }
    for (let mid = 0; mid < s.length; mid++) {
        findMinimumCuts(mid, mid, cutsDp, s);
        findMinimumCuts(mid - 1, mid, cutsDp, s);
    }
    return cutsDp[s.length - 1];
}
function findMinimumCuts(
    startIndex: number,
    endIndex: number,
    cutsDp: number[],
    s: string,
): void {
    for (
        let start = startIndex, end = endIndex;
        start >= 0 && end < s.length && s.charAt(start) == s.charAt(end);
        start--, end++
    ) {
        let newCut: number = start == 0 ? 0 : cutsDp[start - 1] + 1;
        cutsDp[end] = Math.min(cutsDp[end], newCut);
    }
}

File: Editorial/132_Palindrome_Partitioning_II/Go/6_Expand_Around_the_Center.go
func minCut(s string) int {
    cutsDp := make([]int, len(s))
    for i := 1; i < len(s); i++ {
        cutsDp[i] = i
    }
    for mid := 0; mid < len(s); mid++ {
        findMinimumCuts(mid, mid, cutsDp, s)
        findMinimumCuts(mid-1, mid, cutsDp, s)
    }
    return cutsDp[len(s)-1]
}

func findMinimumCuts(startIndex, endIndex int, cutsDp []int, s string) {
    for start, end := startIndex, endIndex; start >= 0 && end < len(s) && s[start] == s[end]; start, end = start-1, end+1 {
        var newCut int
        if start == 0 {
            newCut = 0
        } else {
            newCut = cutsDp[start-1] + 1
        }
        if cutsDp[end] > newCut {
            cutsDp[end] = newCut
        }
    }
}
